### DISCUSSION TITLE
EVE_CFAF800480Ex_050SC_A2 CS and PDN pins cannot be changed on ESP32?

### DISCUSSION THREAD

**Discussion by jwestes** [RudolphRiedel/FT800-FT813#62] [Q&A] [answered]
I have tried moving the CS and PDN pins by add a #define at the top of EVE,h for all 5 pins, see below.  I have tried multiple pin combinations, but the only way I can get anything to display on the screen is to use the defaults of 13 and 12 for CS and PDN.  There are 2 issues with that.  

1. I want to have a J-TAG header on the board which uses pins 12-15 
2. I cannot upload code to the ESP32 with the display connected on pins 12 and 13.  I have to disconnect the pins, upload the code and then reset the board, which also makes debugging difficult.

Is there something I am missing?

#define EVE_CS      32
#define EVE_PDN     27
#define EVE_SCK     18
#define EVE_MISO    19
#define EVE_MOSI    23

**Comment by jwestes** [comment, marked as answer]
I figured it out.  I added defines in the project properties and that fixed it.

**Comment by RudolphRrr** [comment]
Yes, this is maybe a little akward to use right now but it should be a lot easier now that I moved the target definitions out of EVE_target.h.
The idea was that you would adapt EVE_target.h to what you need, as well as EVE_config.h needed to be changed to the module you wanted to use.
And now you need to define the module in the project settings and can overwrite the settings from EVE_target_xxx.h in the project settings as well.

What I should do now that the target definitions all have their own much smaller files is to change this:
#if !defined (EVE_CS)
    #define EVE_CS      GPIO_NUM_13
    #define EVE_PDN     GPIO_NUM_12
    #define EVE_SCK     GPIO_NUM_18
    #define EVE_MISO    GPIO_NUM_19
    #define EVE_MOSI    GPIO_NUM_23
#endif

To this:
#if !defined (EVE_CS)
    #define EVE_CS      GPIO_NUM_13
#endif

#if !defined (EVE_PDN)
    #define EVE_PDN     GPIO_NUM_12
#endif

To allow more atomic adaptions thru the project properties.

And the goal for this is of course that you do not need to change any of the files so you do not have to do it again when you check these out again.
Oh yes, while doing this I also will throw in a comment that if you need to change the settings you can do it from the project settings.

**Reply by jwestes** [reply]
That all sounds good.  While you are at it.  I had to add the following to the top of eve.h just to get it to compile.

#if defined (ARDUINO)
#include <Arduino.h>
#endif

I also found that touch was not working on the display I have until I commented out #define EVE_HAS_GT911 in the config section.  I have a CFAF800480E1050 from 2021.  I don't see anywhere if it is an A2 or not though.

#if defined (EVE_CFAF800480Ex_050SC_A2)
#define Resolution_800x480

#define EVE_PCLK (1L) /* 1 = use second PLL for pixel-clock in BT817 / BT818 */
#define EVE_PCLK_FREQ (30000000L) /* EVE_PCLK needs to be set to 1 for this to take effect */
#define EVE_PCLKPOL (1L)
#define EVE_SWIZZLE (0L)
#define EVE_CSPREAD (0L)
#define EVE_HAS_CRYSTAL
#define EVE_GEN 4
//#define EVE_HAS_GT911 /* special treatment required for out-of-spec touch-controller */
#endif

**Reply by RudolphRiedel** [reply]
Hmm, I just took the Arduino/PlatformIO example from my repository and tried to compile it.
I had to replace RESTORE_CONTEXT() with DL_RESTORE_CONTEXT as I removed the macro and replaced it with a define and did not update any of the examples so far.
But after that the example just builds across all Arduino boards:
uno                        SUCCESS   00:00:01.560
avr_pro                    SUCCESS   00:00:01.522
nano328                    SUCCESS   00:00:01.503
mega2560                   SUCCESS   00:00:01.541
adafruit_metro_m4          SUCCESS   00:00:04.505
samd21_m0-mini             SUCCESS   00:00:02.828
ESP32                      SUCCESS   00:00:08.860
ESP8266_d1_mini            SUCCESS   00:00:06.171
nucleo_f446re              SUCCESS   00:00:06.500
teensy41                   SUCCESS   00:00:03.380
adafruit_feather_nrf52840  SUCCESS   00:00:04.861
bbcmicrobit_v2             SUCCESS   00:00:02.060
teensy35                   SUCCESS   00:00:02.988
xmc1100_xmc2go             SUCCESS   00:00:03.677

The #include <Arduino.h> is in the EVE_target_Arduino_xxx.h file which is included from EVE_target.h and which is the very first that EVE.h includes.


And in regards of the issue with the CFAF800480E1050 I have to check with Crystalfontz, that either was a mistake or I need to put in an second setup.

**Reply by jwestes** [reply]
Strange!  I just commented out the include that I added and now it works.  No idea what that was all about.

**Reply by RudolphRiedel** [reply]
Im am just checking the different variants of the CFAF800480E... from Crystalfontz and unfortunately there is an inconsistency.
In the newer datasheet from 2022-08-30 for the CFAF800480E1-050SC-A2 there is this:
"Capacitive Touch Driver: presents as FT5316 (display actually has a GT911, but in order to
preserve backwards compatibility, it presents as the FT5316)"
In the 2021-07-06 datasheet however it states: "Capacitive Touch Driver: GT911"
The CFAF800480E1-050SC-A1-1 was using a GT911.
The CFAF800480E2-050SC-A2 also is using a GT911.

So, thank you for pointing this out!

**Comment by jwestes** [comment]
BTW - Here are the touch calibrations for the CFAF800480E1_050SC if you want to include them in tft.c

#if defined (EVE_CFAF800480Ex_050SC_A2)
    EVE_memWrite32(REG_TOUCH_TRANSFORM_A, 0x00010603);
    EVE_memWrite32(REG_TOUCH_TRANSFORM_B, 0x0000007A);
    EVE_memWrite32(REG_TOUCH_TRANSFORM_C, 0xFFFF0525);
    EVE_memWrite32(REG_TOUCH_TRANSFORM_D, 0x0000060B);
    EVE_memWrite32(REG_TOUCH_TRANSFORM_E, 0x00010E13);
    EVE_memWrite32(REG_TOUCH_TRANSFORM_F, 0xFFE8813F);
#endif
